src/go/k8s/pkg/resources/statefulset.go

```diff
@@ -878,6 +883,83 @@ func (r *StatefulSetResource) getPorts() []corev1.ContainerPort {
 		return ports
 	}
 
+	ports = append(ports, r.GetPortsForListenersInAdditionalConfig()...)
+
+	return ports
+}
+
+// AdditionalListenersEnvVars returns the env var containing the additioanl listeners specified in additionalConfiguration.
+func (r *StatefulSetResource) AdditionalListenersEnvVars() []corev1.EnvVar {
+	if len(r.pandaCluster.Spec.AdditionalConfiguration) == 0 {
+		return nil
+	}
+
+	cfg := map[string]string{}
+	for _, k := range additionalListenerCfgNames {
+		if v, found := r.pandaCluster.Spec.AdditionalConfiguration[k]; found {
+			cfg[k] = v
+		}
+	}
+	if len(cfg) == 0 {
+		return nil
+	}
+	jsonStr, err := json.Marshal(cfg)
+	if err != nil {
+		return nil
+	}
+	return []corev1.EnvVar{{
+		Name:  "ADDITIONAL_LISTENERS",
+		Value: string(jsonStr),
+	}}
+}
+
+// GetPortsForListenersInAdditionalConfig gets the ports for the additional listeners and advertised APIs set in addtionalConfiguration.
+// - redpanda.kafka_api
+// - redpanda.advertised_kafka_api
+// - pandaproxy.pandaproxy_api
+// - pandaproxy.advertised_pandaproxy_api
+// example: redpanda.kafka_api: "[{'name':'private-link','address':'0.0.0.0','port':39002}]"
+// example: redpanda.advertised_kafka_api: "[{'name':'private-link','address':'{{ .Index }}-f415bda0-{{ .HostIP | sha256sum | substr 0 7 }}.clb2jkpb06k16j807u20.byoc.ign.cloud.redpanda.com','port': 'port': {{30092 | add .Index}}}]"
```

Comment:
NIT: If possible please redact or change our cloud DNS names in public repository.

Reply:
Done.